### ISSUE TITLE
NPE fixed in InFilterConstraint.apply()

### ISSUE THREAD

**Issue by Quix0r** [primefaces/primefaces#3012]
## 1) Environment
- PrimeFaces version: **6.2-master** and **6.1 (stable)**
- Does it work on the newest released PrimeFaces version? **No**
- Does it work on the newest sources in GitHub? **No**
- Application server + version: **Payara Server  4.1.2.173 #badassfish (build 24)]]**
- Affected browsers: **independent**

## 2) Expected behavior

The below code should allow the user to filter non-assigned branch offices (custom stuff, basically it is an entity that can be assigned (means set) in an other entity `employee`.

...

## 3) Actual behavior

The code throws an NPE:
````
[2017-11-27T22:31:11.802+0100] [Payara 4.1] [SEVERE] [] [javax.enterprise.resource.webcontainer.jsf.context] [tid: _ThreadID=31
 _ThreadName=http-thread-pool::http-listener-1(5)] [timeMillis: 1511818271802] [levelValue: 1000] [[
  java.lang.NullPointerException
        at org.primefaces.model.filter.InFilterConstraint.applies(InFilterConstraint.java:47)
        at org.primefaces.component.datatable.feature.FilterFeature.filter(FilterFeature.java:194)
        at org.primefaces.component.datatable.feature.FilterFeature.encode(FilterFeature.java:125)
        at org.primefaces.component.datatable.DataTableRenderer.encodeEnd(DataTableRenderer.java:85)
        at javax.faces.component.UIComponentBase.encodeEnd(UIComponentBase.java:949)
        at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1912)
        at com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:638)
        at com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:183)
        at org.primefaces.component.api.UIData.visitTree(UIData.java:827)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIForm.visitTree(UIForm.java:395)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1747)
        at com.sun.faces.context.PartialViewContextImpl.processComponents(PartialViewContextImpl.java:423)
        at com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:342)
        at org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:57)
        at javax.faces.component.UIViewRoot.encodeChildren(UIViewRoot.java:1124)
        at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1905)
        at com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:465)
        at com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:194)
        at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:151)
        at com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:126)
        at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:100)
        at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:223)
        at javax.faces.webapp.FacesServlet.service(FacesServlet.java:671)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1606)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:338)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.jboss.weld.probe.ProbeFilter$FilterAction.proceed(ProbeFilter.java:329)
        at org.jboss.weld.probe.ProbeFilter$FilterAction.proceed(ProbeFilter.java:312)
        at org.jboss.weld.probe.InvocationMonitor$Action.perform(InvocationMonitor.java:157)
        at org.jboss.weld.probe.ProbeFilter$FilterAction.doFilter(ProbeFilter.java:338)
        at org.jboss.weld.probe.ProbeFilter.embedInfoSnippet(ProbeFilter.java:188)
        at org.jboss.weld.probe.ProbeFilter.doFilter(ProbeFilter.java:176)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:250)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:250)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:256)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:654)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:593)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:155)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:480)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:180)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:539)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573)
        at java.lang.Thread.run(Thread.java:748)
]]
````

..

## 4) Steps to reproduce

- There is now an example app (see comments), do the usual steps: Download+import it into NetBeans (8.1 at least), build WAR.
- Copy shipped Primefaces-6.1.jar to your `~/personal_domain/lib/` folder (or equivalent on other OSes)
- deploy the before created WAR and launch it (e.g. use the **Run** target)
- Now unselect all (bug?) and the list should remain
- select 1/2 and the list should change
- select "None selected" and the list goes empty but a NPE is logged.
- stop your app server
- build PF from this PR and replace 6.1 with the just build PF version
- repeat above steps and the bug should be gone.
.. 

## 5) Sample XHTML

Relevant snippet:
````
<f:facet name="filter">
	<p:selectCheckboxMenu
		filter="true"
		filterMatchMode="contains"
		label="#{msg.LABEL_BRANCH_OFFICES}"
		onchange="PF('employeeList').filter()"
		updateLabel="true"
		title="#{msg.FILTER_BY_MULTIPLE_EMPLOYEES_TITLE}"
		>
		<f:converter converterId="BranchOfficeConverter" />
		<f:selectItem itemValue="#{null}" itemLabel="#{msg.NONE_SELECTED}" />
		<f:selectItems value="#{branchOfficeController.allBranchOffices()}" var="branchOffice" itemValue="#{branchOffice}" itemLabel="#{beanHelper.renderBranchOffice(branchOffice)}" />
	</p:selectCheckboxMenu>
</f:facet>
````

..

## 6) Sample bean

..

Please see #3011 for my code changes.

**Comment by Quix0r** [comment]
Coming up with a small-repeatable and easy to use example code.

**Comment by tandraschko** [comment]
IMO we can skip a example, a null check doesnt hurt.

**Comment by Quix0r** [comment]
Already working on it. ;-)

**Comment by Quix0r** [comment]
Done: Build with NetBeans+ant:
https://cloud.mxchange.org/index.php/s/149vSQYmDeENiuY/download?path=%2FPrimefaces&files=NpeInFilterConstraint.zip

**Comment by Quix0r** [comment]
Updated steps to reproduce due to the example app.

**Comment by tandraschko** [comment]
@mertsincan This is just a little null check, maybe you can put into elite as the user asked for it.

**Comment by Quix0r** [comment]
Has it also been included in 6.1? I mean, it is a small trivial yet important change, everyone should benefit from it.

**Comment by Quix0r** [comment]
Sadly, it didn't make it in, not even as a patch with an increased patch level, like 6.1.1 or so. I still have to and will wait for 6.2.
